dwarf-rust-experimental/rust/src/main.rs

```diff
@@ -0,0 +1,502 @@
+  }
+  if filename.contains("main.rs") || filename.contains("lib.rs") {
+    score += 50; // Prefer entry point files
+  }
+  if filename.contains("mod.rs") && !filename.contains("intrinsics") {
+    score += 30; // Prefer module files (but not intrinsics)
+  }
+
+  // Strategy 3: Avoid generated/internal files
+  if filename.contains("intrinsics") || filename.contains("panic") {
+    score -= 150; // Avoid compiler intrinsics and panic handlers
+  }
+  if filename.contains("macros.rs") {
+    score -= 100; // Macro definitions are often less useful for debugging
+  }
+  if filename.contains("impls.rs") || filename.contains("builders.rs") {
+    score -= 80; // Implementation details vs user code
+  }
+
+  // Strategy 4: Prefer specific line numbers over line 0
+  if line_number == 0 {
+    score -= 200; // Line 0 is often a catch-all or generated
+  } else if line_number < 10 {
+    score -= 50; // Very early lines might be imports/boilerplate
+  }
+
+  // Strategy 5: Prefer shorter, simpler paths (user code is typically closer to root)
+  if let Some(dir) = directory {
+    let path_depth = dir.matches('/').count();
+    if path_depth > 5 {
+      score -= (path_depth as u32 - 5) * 20; // Penalize deeply nested paths
+    }
+  }
+
+  score
+}
+
+fn run() -> Result<(), Error> {
+  let mut buffer = Vec::new();
+  std::io::stdin()
+    .read_to_end(&mut buffer)
+    .map_err(Error::Io)?;
+  let slice = EndianSlice::new(buffer.as_slice(), LittleEndian);
+  let result = extract_source_info(slice)?;
+  serde_json::to_writer(std::io::stdout(), &result).map_err(Error::Json)?;
+  Ok(())
+}
+
+fn main() {
+  run().unwrap_or_else(|err| {
+    let error_doc = SourceResult {
+      error: Some(err.to_string()),
+      units: None,
+      lines: None,
+      functions: None,
+    };
+    serde_json::to_writer(std::io::stdout(), &error_doc).unwrap_or_else(|err| {
+      eprintln!("Error: {}", err);
+      std::process::exit(2);
+    });
+    std::process::exit(1);
+  });
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+
+  #[test]
+  pub fn test_parse() {
+    // read a file in
+    let src = std::fs::read("./src/count_vowels.wasm").expect("could not read wasm file");
+    let actual = extract_source_info(EndianSlice::new(src.as_slice(), LittleEndian))
+      .expect("extract_soruce_info failed");
+
+    // Optionally write the expected test data when REGENERATE_TEST_DATA env var is set
+    if std::env::var("REGENERATE_TEST_DATA").is_ok() {
+      std::fs::write(
+        "./src/count_vowels.wasm.json",
+        serde_json::to_string_pretty(&actual).expect("json failed"),
+      )
+      .unwrap();
+      println!("Regenerated test data in ./src/count_vowels.wasm.json");
+    }
+
+    // read count_vowels.rs.wasm.json
+    let expected: SourceResult = serde_json::from_reader(
+      std::fs::File::open("./src/count_vowels.wasm.json").expect("could not read json file"),
+    )
+    .expect("json failed");
+    assert_eq!(actual, expected);
```

Comment:
this looks to be the library to use in Rust:
https://github.com/mitsuhiko/insta

That said `count_vowels.wasm.json` is too big, there is no chance we can effectively inspect it... we probably need something different than an approval.

Reply:
The file is not there to be inspected line by line.  It's there to demonstrate an example of the output that this module produces.  Instead of reading rust code to figure out how to implement the java parser for it, I just look at the example data file it produces to understand it's structure so I can write the jackson bindings.

Reply:
I see, can we have a super minimal one?